src/org/omegat/util/gui/UIDesignManager.java

```diff
@@ -69,16 +72,84 @@
  * @author Alex Buloichik
  * @author Yu Tang
  * @author Aaron Madlon-Kay
+ * @author Hiroshi Miura
  */
 public final class UIDesignManager {
 
+    public static final String THEME_DEFAULT_ICON = "thumb_default.png";
+    public static final String THEME_UNKNOWN_ICON = "thumb_unknown.png";
+
+    private static final Map<String, URL> themeImages = new HashMap<>();
+    private static final Map<String, IThemeDesigner> themeInitilizer = new HashMap<>();
+
+    static {
+        themeImages.put(Preferences.THEME_DEFAULT, ResourcesUtil.getResourceURL(THEME_DEFAULT_ICON));
+        for (UIManager.LookAndFeelInfo info: UIManager.getInstalledLookAndFeels()) {
+            String key = info.toString();
+            String icon = "thumb_" + info.getName().toLowerCase().replace("/", "_").replace("+", "_") + ".png";
+            URL thumbURL = ResourcesUtil.getResourceURL(icon);
+            if (thumbURL == null) thumbURL = ResourcesUtil.getResourceURL(THEME_UNKNOWN_ICON);
+            themeImages.put(key, thumbURL);
+        }
+    }
+
     private UIDesignManager() {
     }
 
+    public static Map<String, String> getThemes() {
+        Map<String, String> themes = new TreeMap<>();
+        themes.put("Default", OStrings.getString("PREFS_THEME_NAME_DEFAULT"));
+        UIManager.LookAndFeelInfo[] lafInfo = UIManager.getInstalledLookAndFeels();
+        for (UIManager.LookAndFeelInfo lookAndFeelInfo : lafInfo) {
+            String name = lookAndFeelInfo.getName();
+            String key = lookAndFeelInfo.toString();
+            themes.put(key, name);
+        }
+        return themes;
+    }
+
+    @SuppressWarnings("unused")
+    public static void registerTheme(IThemeDesigner designer) {
+        registerTheme(designer.getName(), designer.getClassName(), designer.getIcon());
+        themeInitilizer.put(designer.getName(), designer);
+    }
+
+    public static void registerTheme(String name, String className, URL image) {
+        UIManager.LookAndFeelInfo info = new UIManager.LookAndFeelInfo(name, className);
+        UIManager.installLookAndFeel(info);
+        if (image != null) {
+            String key = info.toString();
+            themeImages.put(key, image);
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public static void registerTheme(String name, String className) {
+        registerTheme(name, className, ResourcesUtil.getResourceURL(THEME_UNKNOWN_ICON));
+    }
+
+    public static Set<String> getThemeKeySet() {
+        return themeImages.keySet();
+    }
+
+    public static URL getThemeImage(String key) {
+        return themeImages.getOrDefault(key, ResourcesUtil.getResourceURL(THEME_UNKNOWN_ICON));
+    }
+
+    public static String getThemeClassName(String key) {
```

Comment:
I don't understand why using `UIManager.LookAndFeelInfo#toString()` as the key is better than simply using the classname as the key. If we used the classname as the key then we wouldn't even need this particular lookup.

Reply:
It is because LaF class can have multiple design. So classname may not be unique against theme.
toString include a name and classname that is unique.

Reply:
Aha. Well that is a good reason, thanks.

Reply:
Actually I'm still confused though: to apply the theme we call `UIManager.setLookAndFeel(classname)`. If the classname is not enough to uniquely identify the theme, then how do different "designs" of the same theme take effect? Are there any examples of such themes in the real world?

I see in your sample plugin

https://github.com/miurahr/omegat-flatlaf/blob/7922c2cd57cc6315bf5b96b453f37ece6e7607f8/src/main/java/tokyo/northside/omegat/laf/CustomLookAndFeel.java#L35-L39

every theme has a different implementing class.

Even if it's hypothetically true that a single LAF class could implement multiple different "themes", unless such usage is common in the real world, I would prefer to require that each theme be a separate class. That would simplify a lot of our API design.

Reply:
I've pushed a case that a theme use same LaF class with other, but provide different theme design (round corner of buttons, and triangle arrow)
https://github.com/miurahr/omegat-flatlaf/blob/master/src/main/java/tokyo/northside/omegat/laf/CustomLookAndFeel.java#L42

Reply:
Here is a plugin jar that has a "Flat Round Dark" and "FlatLaf Dark" theme which both use "FlatDarkLaf" class.

[omegat-flatlaf-0.0.1.jar.gz](https://github.com/omegat-org/omegat/files/6592018/omegat-flatlaf-0.0.1.jar.gz)

Reply:
Here is another plugin examples.
https://github.com/miurahr/omegat-themes
[omegat-themes-0.0.1.jar.gz](https://github.com/omegat-org/omegat/files/6598610/omegat-themes-0.0.1.jar.gz)

Reply:
Thanks.

However I it looks like the "multiple themes from a single LAF class" is simply a convention that you invented, not something that is widely practiced in the real world, or that is somehow required by external concerns (e.g. the LAF API encourages it somehow).

I was hoping for something like your idea here: https://github.com/omegat-org/omegat/pull/100#discussion_r645269366

I would much prefer that each theme is required to be encapsulated in its own LAF class, and not require additional configuration.

Is there a reason that can't be done? Let's continue the discussion either in this thread or in https://github.com/omegat-org/omegat/pull/100#discussion_r645269366.